src/metabase/search/postgres/index.clj

```diff
@@ -0,0 +1,124 @@
+
+(def ^:private tsv-language "english")
+
+(defn- exists? [table-name]
+  (t2/exists? :information_schema.tables :table_name (name table-name)))
+
+(defn- drop-table! [table-name]
+  (boolean
+   (when (exists? table-name)
+     (t2/query (sql.helpers/drop-table table-name)))))
+
+(defn- rename-table! [old new]
+  (when (and (exists? old) (not (exists? new)))
+    (-> (sql.helpers/alter-table old)
+        (sql.helpers/rename-table new)
+        t2/query)))
+
+(defn create-pending!
+  "Create a non-active search index table."
+  []
+  (when (not @reindexing?)
+    (when-not (exists? pending-table)
+      (-> (sql.helpers/create-table pending-table)
+          (sql.helpers/with-columns
+            [[:id :bigint [:primary-key] [:raw "GENERATED BY DEFAULT AS IDENTITY"]]
+             ;; entity
+             [:model_id :int :not-null]
+             [:model [:varchar 254] :not-null] ;; TODO find the right size
+             ;; search
+             [:search_vector :tsvector :not-null]
+             ;; scoring related
+             [:model_rank :int :not-null]
+             ;; permission related entities
+             [:collection_id :int]
+             [:database_id :int]
+             [:table_id :int]
+             ;; filter related
+             [:archived :boolean]])
+          t2/query)
+
+      (t2/query
+       (format "CREATE INDEX IF NOT EXISTS %s_tsvector_idx ON %s USING gin (search_vector)"
+               (str/replace (str (name active-table) "_" (random-uuid)) #"-" "_")
+               (name pending-table))))
+    (reset! reindexing? true)))
+
+(defn activate-pending!
+  "Make the pending index active, if it exists. Returns true if it did so."
+  []
+  ;; ... just in case it wasn't cleaned up last time.
+  (drop-table! retired-table)
+  (when (exists? pending-table)
+    (t2/with-transaction [_conn]
+      (rename-table! active-table retired-table)
+      (rename-table! pending-table active-table))
+    (reset! reindexing? false)
+    (drop-table! retired-table)
+    true))
+
+(defn update!
+  "Create or update the given search index trny"
+  [entity]
+  (let [entry (-> entity
+                  (select-keys
+                   [:model
+                    :model_rank
+                    :collection_id
+                    :database_id
+                    :table_id
+                    :archived])
+                  (assoc
+                   :model_id      (:id entity)
+                   :search_vector [:to_tsvector
+                                   [:inline tsv-language]
+                                   [:cast
+                                    (:searchable_text entity)
+                                    :text]]))]
+    (when @initialized?
+      (t2/insert! active-table entry))
+    (when @reindexing?
+      (t2/insert! pending-table entry))))
+
+(def ^:private ts-query
+  [:raw
+   "search_vector @@ websearch_to_tsquery('"
+   tsv-language "', "
+   [:param :search-term] ")"])
+
+(def search-query
```

Comment:
You can use `:lift` in a raw clause for params.
```clojure
(defn search-query
  "Query fragment for all models corresponding to a query paramter `:search-term`."
  [search-term]
  {:select [:model_id :model]
   :from   [active-table]
   :where  [:raw
            "search_vector @@ websearch_to_tsquery('"
            tsv-language "', "
            [:lift search-term] ")"]})
```

I think this is cleaner and we don't have to remember about that `search-term` params every where we use it.

Reply:
TIL!